### ISSUE TITLE
Proposal: String enums

### ISSUE THREAD

**Issue by jbondc** [Microsoft/TypeScript#3192]
A string enum is inferred by looking at the constant value of the initializer of its first member:

```
enum stringsInferred {
   a = "a",
   b, // "b"
   c  // "c"
}
```

Or it is declared using a string index:

```
enum stringsDeclared {
   [prop: string] : string;
   a, // "a"
   B, // "B"
   c  // "c"
}
```

The auto initialized value on an enum is its property name. 
## Inferred enum special case

```
enum stringsLowerCase {
   Shoes = "shoes", // If the first member has lowercase chars, then lowercase all other member values
   Boots, // "boots"
   Feet  // "feet"
}
```

**Comment by mariusschulz** [comment]
:+1:

We should definitely be able to use enums with string values. Even a `const enum` with compile-time string literals that are safe to inline would be helpful already, mainly for …

- restricting variables to a set of known string values, and
- avoiding misspellings.

I'm not sure the automagical upper- and lowercasing described underneath *Inferred enum special cases* is such a good idea, though. I'd rather spell all cases exactly like they should be stringified:

```typescript
enum stringsUpperCase {
   Shoes = "SHOES",  // "SHOES"
   BOOTS,            // "BOOTS"
   Feet              // "Feet"
}
```

**Comment by jbondc** [comment]
@mariusschulz You could always use the declared form (which doesn't infer anything):
```
enum stringsUpperCase {
    [prop: string] : string;
   Shoes = "SHOES",  // "SHOES"
   BOOTS,            // "BOOTS"
   Feet              // "Feet"
}
```
The automatic lower-casing comes from @SaschaNaz :
https://typescript.codeplex.com/discussions/549207

I'll admit that maybe the uppercase one isn't needed.

**Comment by impinball** [comment]
Can this be generalized to other types (with the constraint that all types must be initialized - none can be inferred)?

**Comment by jbondc** [comment]
@impinball It can, more elaborate examples in #3105

Just wanted to keep this discussion focused on the enum with values of type string.
Seems a bit redundant to not allow inferring the type:
```
// (a)
enum strings {
   Shoes = "Shoes",  // "Shoes"
   BOOTS,            // "BOOTS"
   Feet              // "Feet"
}
// (b)
enum strings {
   [prop: string] : string;
   Shoes,           // "Shoes"
   BOOTS,            // "BOOTS"
   Feet              // "Feet"
}
```
Both work but find (a) more visually appealing.

**Comment by impinball** [comment]
@jbondc 

I agree, and this could be the other special case (where value inference already exists for numbers).

**Comment by impinball** [comment]
@jbondc 

And I feel this syntax is a little repetitive in the `[prop: string]: string` part. Couldn't the syntax be modified to have the type after the enum name, instead of in the body? Maybe this syntax, to borrow your example?

```ts
enum strings : string {
   Shoes,           // "Shoes"
   BOOTS,            // "BOOTS"
   Feet              // "Feet"
}
```

**Comment by jbondc** [comment]
Certainly could and it's a neat looking syntax. 

Went with ``` [prop: string]: string``` because:
(a) Same looking syntax as with interfaces
(b) The property index adds some expressiveness so you could have things like:
```
type UpperFirst = /[A-Z][A-Za-z]+/;
enum strings {
   [prop: UpperFirst] : string;
   Shoes,           // "Shoes"
   BOOTS,            // "BOOTS"
   Feet              // "Feet"
}
```
My gamble is it has some usefulness if you want to create a library that has some opinionated styles to it :grinning: 
So less about type safety but using types as a mechanism to enforce constraints.

**Comment by jbondc** [comment]
Not allowing anyone to extend your enum:
```
enum strings {
   [prop: "Shoes" | "BOOTS" | "Feet"] : string;
   Shoes,
   BOOTS,
   Feet
}
// Or something like this
enum strings {
   [prop: keys(this)] : string;
   Shoes,
   BOOTS,
   Feet
}
```
These same patterns naturally apply to interfaces too.

**Comment by impinball** [comment]
@jbondc Any advantages to that over my [recent proposal](https://github.com/Microsoft/TypeScript/issues/1206#issuecomment-121926668)? I don't think that regex-based solution is really that helpful in practice. All that regex does there is enforce type names, something better left to a linter as opposed to the main compiler. And it looks a little out of place compared to the enum entries. And [as you said in the other thread](https://github.com/Microsoft/TypeScript/issues/1206#issuecomment-121899504), it will complicate the parser and compiler a little, likely more than it should.

And as for extending enums, that's not very good practice, anyways. And it's generally not possible among other TypeScript users. Mine allows for such extensions, but you can only truly extend them in the same module or private namespace declaration, unless you're writing a definition file. That's because of the [module wrapper emitted around every namespace declaration](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#106-code-generation). So, for most purposes, you can't really extend them in the first place.

**Comment by jbondc** [comment]
Always hard to evaluate usefulness of things that don't exist so no strong opinion about it. A subjective argument is that learning one pattern for ```interface a { [x: string]: string }``` and ```enum b { [x: string]: string}``` adds up to perceived simplicity across the language. Initially didn't like the ```{ [x: a]: string }``` syntax but once I adopted it, it fits naturally object looking declarations. It's not like you declare enums every day so a few more characters for expressiveness wouldn't seem to hurt.

**Comment by lazutkin** [comment]
:+1:  for string (const) enums

**Comment by rotemdan** [comment]
My current workaround looks something like this (this is a real-world, copy-and-paste code example):
```ts
class Task {
  ..
  static State = { 
    Unstarted: "Unstarted", 
    Started: "Started", 
    Completed: "Completed", 
    Faulted: "Faulted", 
    Canceled: "Canceled"
  }
}

let task = new Task();
..
if (task.state === Task.State.Completed)
  console.log("Done!");
```
This also gives a nice encapsulation into the class type that I find appealing (through the `static` modifier). 

Replicating this through `enum` would require several different enhancements. I will consider posting a separate proposal for class encapsulated enums if that hasn't been proposed before.

I think having string enums would be very useful (maybe "necessary" should be the right word here). Automatic inferring of things like letters would be problematic because of locale and language issues (every language would have its own letters and ordering). Automatic lowercasing/uppercasing would also be impacted from similar issues (every language has different lower/uppercasing rules), and seems to me like a minor detail/enhancement at this point.

**Comment by mariusschulz** [comment]
Maybe the `string` keyword would lend itself to be included in the declaration:

```ts
string enum PromiseState {
    Pending,
    Fulfilled,
    Rejected
}
```

In the above example, the string values of the cases would equal their names. They could be overwritten like this:

```ts
string enum PromiseState {
    Pending = "pending",
    Fulfilled = "fulfilled",
    Rejected = "rejected"
}
```

**Comment by rotemdan** [comment]
This may turn out a bit long for a `const enum` that is also exported:
```ts
export const string enum PromiseState {
    Pending,
    Fulfilled,
    Rejected
}
```
So maybe an alternative using some sort of a psuedo-generic type parameter?
```ts
export const enum PromiseState<string> {
    Pending,
    Fulfilled,
    Rejected
}
```
Due to [issues with different languages, alphabets, and locales](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLocaleLowerCase) it would be difficult to apply a correct conversion to lowercase/uppercase in many languages other than English. So just preserving the plain casing and allowing custom ones through `=` seems like a reasonable compromise.

**Comment by jbondc** [comment]
@rotemdan Do you have a test case in mind?

Works fine:
![snap](https://cloud.githubusercontent.com/assets/199519/9688863/ec40b534-52ff-11e5-98ff-d4e4bed76543.png)

Emit is:
```
var bar;
(function (bar) {
    bar["ÉLÉGANT"] = "élégant";
    bar["ΚΟΜΨΌ"] = ""κομψό"";
    bar["NO ISSUE"] = ""no issue"";
})(bar || (bar = {}));
```

**Comment by jbondc** [comment]
There is some usefulness with auto lowercasing enums if you to do case insentive searches in a database for example. Arguably, I think it's a better default. Programmatically some coding conventions say that:
```
module myPhilosophy {
   export enum {
     CamelCaseA,
     CamelCaseB
   }
}
```
is "better". Don't want to get into that... but you get the idea.

**Comment by Gambero81** [comment]
Hi, what is the actual state of this proposal? can we have a roadmap to know when will be implemented in the language? Thanks

**Comment by impinball** [comment]
@Gambero81 See #1206 for a more up-to-date version of this.

**Comment by impinball** [comment]
Can this be closed?

**Comment by jbondc** [comment]
@impinball No, added this here strictly for discussing a string enum & use cases. 
e.g. can you interest two string enums?
```
enum a {
   hello = "hello"
}
enum b {
    world = "world"
}
let c = a & b;
let d = typeof a & typeof b;
```

**Comment by impinball** [comment]
Union and intersection types would already exist with my proposal. You
could achieve the same thing.

**Comment by vitrilo** [comment]
Using of String are really neat feature. In our 20k loc project we still use workarounds.
Apart, using objects as enum values (like in #1206 or http://stackoverflow.com/a/15491832/1224258) are risky - comparison fails after any Enums serialization/de-serialization or deepcopy (angular.copy).

**Comment by basarat** [comment]
Probably better if https://github.com/Microsoft/TypeScript/issues/1003 is done instead

**Comment by impinball** [comment]
@basarat I think both have their uses. A primitive enum (excluding symbols) could be seen as the union of their values in that sense. And also, I believe that should be expanded to also include numbers.

**Comment by impinball** [comment]
I made a relevant response over there on #1003.

**Comment by jbondc** [comment]
@basarat The two are related:
```
const enum listDrinks {
    Coffee = "coffee",
    Tea
}
let drink: listDrinks; // Can have values "coffee" | "tea"
drink = linkDrinks.Coffee;
drink = "beer"; // error
let drinkType: "coffee" | "tea" = "tea";
drinkType = "beer"; // error
```
With enough effort 'Flag me down' ```#4``` can be solved:
http://www.informit.com/articles/article.aspx?p=2425867

Oh and const enums can't have null or undefined values, so hmm... food for thought.

**Comment by impinball** [comment]
@jbondc I tried to include those, but I didn't get a very positive response from it.

**Comment by jbondc** [comment]
@impinball Meant more:
```
const enum numbers {
   a = null, // error
   b = undefined, // error
   one = 1,
}
```
Seems clear this should error (the same for string enums), effectively these are non-nullable types.
```
function foo (): numbers  {
   return null;
}
```

**Comment by vitrilo** [comment]
@basarat Thanks, I've just discovered related #1003, #1295. Great proposals/ideas!
My understanding #1003 (string literal union type) vs #3192 (string enums) are following:
* String enums are simple, straightforward approch
*  #1003 - are more wider idea, which are may be both: multi-purpose (like for type guards) sulution and space for funky/unwanted/corner cases
* Syntax are RADICALLY different (which not obvious on first site):
```javascript
    var var1 = MyEnum.Kind1;
    var var2 = "kind1";
```
  * Readability: IMHO, String enums have better readability. Moreover, in many languages enum types make uppercased to highlight limited possibilities and static nature.
  * IDE support. Tools are all for us! Possible, implementing auto-complete for #1003 (string literal union type) may be significantly harder inside IDEs comparing to  #3192 (string enums). Not because of complexity of union-type concept, also because IDE should introduce in-natural auto-complete inside string literals: "kind1" (should suggest kind2,kind3.. if this is union type).

Apart, from all above, Javascript ecosystem generally encourage wide usage of string literals. As result, like in #1003, #1295 IDEs/Tools will be forced to support special meaning of string literals, like as enum constants/properties names etc, not just meaningless text to output.

**Comment by jtheoof** [comment]
:+1: This feature would be great. 

Right now, the `enums` are a bit counter intuitive from a JavaScript usage perspective. They work well for internal TypeScript code but not so much when code is used in third party apps.

It would be especially useful for bridging the gap between both worlds: 

Example:

```
sdk.ts
--------

export enum Environments {
  SANDBOX,
  PRODUCTION
}

interface ClientOptions {
  id: string;
  secret: string;
  env?: Environments;
}

export class Client {
  constructor(public options: ClientOptions) {}

  hostname(): string {
    let part = 'sandbox';

    if (this.env === Environments.PRODUCTION) {
      part = 'api';
    }

    return `rest.${part}.example.com`;
  }
}

caller.js
----------

var client = new Client({
  id: 'id',
  secret: 'secret',
  env: 'production'
});

```

Right now, `client.hostname()` will fail because `Environments.PRODUCTION === 1`.

**Comment by mrcrowl** [comment]
:+1:

**Comment by olmobrutall** [comment]
My two cents while this gets approved:

Not that you can pretend this feature like this.

```TS
    export enum OrderType {
        Ascending = "Ascending" as any,
        Descending = "Descending" as any,
    }
```

The benefit is that then your API is better documented, asking for an `OrderType` instead of a dirty string. 

The problem is that you need to repeat the identifier for the common case when declaring. 

Also, when using it, you have to cast from `string` -> `any` -> `OrdeType`:


```TS
var orderType: OrderType = "Ascending"; //Error
var orderType: OrderType = "Ascending" as any; //OK
```

**Comment by DanielRosenwasser** [comment]
Here's a way you'll be able to emulate string enums in `typescript@next`. Taking a modified version of Esprima's `Syntax` object:

```ts
function strEnumify<T extends { [prop: string]: "" | string }>(obj: T) {
    return obj;
}

export const Syntax = strEnumify({
    AssignmentExpression: 'AssignmentExpression',
    AssignmentPattern: 'AssignmentPattern',
    ArrayExpression: 'ArrayExpression',
    ArrayPattern: 'ArrayPattern',
    ArrowFunctionExpression: 'ArrowFunctionExpression',
    BlockStatement: 'BlockStatement',
    BinaryExpression: 'BinaryExpression',
    BreakStatement: 'BreakStatement',
    CallExpression: 'CallExpression',
    CatchClause: 'CatchClause',
    ClassBody: 'ClassBody',
    ClassDeclaration: 'ClassDeclaration',
    ClassExpression: 'ClassExpression',
    // ...
    ThisExpression: 'ThisExpression',
    ThrowStatement: 'ThrowStatement',
    TryStatement: 'TryStatement',
    UnaryExpression: 'UnaryExpression',
    UpdateExpression: 'UpdateExpression',
    VariableDeclaration: 'VariableDeclaration',
    VariableDeclarator: 'VariableDeclarator',
    WhileStatement: 'WhileStatement',
    WithStatement: 'WithStatement',
    YieldExpression: 'YieldExpression'
});
```

![image](https://cloud.githubusercontent.com/assets/972891/11854189/1a3ab12c-a3f7-11e5-9e27-2fba97ffe226.png)

(@ariya)

The problem is that you can't do something like the following:

```ts
let kind = Syntax.VariableDeclaration;
kind = Syntax.WithStatement;
```

**Comment by plantain-00** [comment]
```typescript
export type ResponseType = "json" | "others";

export const responseType = {
    json: <ResponseType>"json",
    others: <ResponseType>"others",
};
```
Like this, you can get a type, and tips when typing from `responseType`.

**Comment by DanielRosenwasser** [comment]
When doing that, the `json` property has type `"json" | "others"` if that's fine with you.

**Comment by vitrilo** [comment]
@olmobrutall thanks for perfect workaround! Well serialized/deserialized, apart from unreadable numeric enums inside json. 
Below working even in v1.4: 

      enum OrderType { Ascending = <any>"Ascending", Descending = <any>"Descending" }

**Comment by isiahmeadows** [comment]
@DanielRosenwasser I would generally prefer to use const numeric enums where possible (a JS-to-ESTree AST parser couldn't for spec reasons) for that just for memory and performance reasons (both reduced for free). But I could see merit in a string enum to hold error messages like what [jshint uses](https://github.com/jshint/jshint/blob/master/src/messages.js).

**Comment by Gambero81** [comment]
@olmobrutall @DanielRosenwasser @vitrilo there workaround are good but a good feature of string enum should be to inlining string const enum without emit js as done with standard const enum, unfortunally for this porpouse all these workaround doesn't work

**Comment by olmobrutall** [comment]
@Gambero81, of course having native support from the language will be the best solution, simplifying the declaration (no repetition for the simple cases), and the usage (no casting). 

About const enums, we already have the syntax and the concept for normal enums, it will make sense just to extrapolate the same concepts for string enums. 


```Ts
export const enum Color extends string{
   red,
   blue,
   darkViolet= "dark-violet"
}
```

**Comment by hgoebl** [comment]
:+1: current project: accessing REST service with generated TypeScript interfaces and enums. Enums are not usable, since service needs strings in JSON payload.

Is there a way to serialize / deserialize JSON and convert enum instances on-the-fly?

**Comment by olmobrutall** [comment]
I've just found this today: https://blog.wearewizards.io/flow-and-typescript-part-2-typescript

TLDR; Typescript is better than flow but has no string enums :-1: 

> The only issues we have encountered so far (and would love help to resolve!) are:
* webpack building the previous version: adding an empty line and saving a file re-triggers a compilation but can be frustrating
* no string enums: we want to keep human readable enum members but TypeScript only supports int enums

Any news on this issue? Any change this gets implemented.

**Comment by Gambero81** [comment]
I agree.. string enum is a MUST!! (in particular const string enum)

still hope to show it soon...

**Comment by Gambero81** [comment]
I agree.. string enum is a MUST!! (in particular const string enum)

still hope to show it soon...

**Comment by Gambero81** [comment]
I agree.. string enum is a MUST!! (in particular const string enum)

still hope to show it soon...

**Comment by sampsyo** [comment]
I just read that the new TypeScript 1.8 beta has "string literal types": https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#string-literal-types

That seems to solve exactly this use case, right?

    type Order = "Ascending" | "Descending";

**Comment by Gambero81** [comment]
string literal is not string enum..

**Comment by olmobrutall** [comment]
Wow didn't know that. With autocompletion they can be a even better alrernative! 
```TS
column.order = //autocompletion here please!
               //"Ascending"
               //"Descending"
```
Thanks TS Team!

**Comment by basarat** [comment]
@olmobrutall I have two issues created for your tracking :rose: https://github.com/Microsoft/TypeScript/issues/5602 https://github.com/Microsoft/TypeScript/issues/6705

**Comment by dsherret** [comment]
String enums can be mimicked by merging a string literal type and module this way:

```typescript
type MyStringEnum = "member1" | "member2";

module MyStringEnum {
    export const Member1: MyStringEnum = "member1";
    export const Member2: MyStringEnum = "member2";
}
```

[Full example](http://stackoverflow.com/a/35257367/188246).

It seems it's only worth implementing `enum MyStringEnum { myMember = "my member" }` for syntactic sugar now.

**Comment by olmobrutall** [comment]
@dsherret nice point!

**Comment by isiahmeadows** [comment]
@dsherret `s/module/namespace/g`. `module` in that context is deprecated, and it has been for a while. :smile:

**Comment by dsherret** [comment]
I forgot declaration merging worked with a variable. This is better:

```typescript
type MyStringEnum = "member1" | "member2";

const MyStringEnum = {
    Member1: "member1" as MyStringEnum,
    Member2: "member2" as MyStringEnum
};
```

[Updated example](http://stackoverflow.com/a/35257367/188246).

**Comment by amcdnl** [comment]
Clever hack @dsherret but why is this not implemented yet? +1 for string enums!

**Comment by isiahmeadows** [comment]
@amcdnl I feel the more general #1206 is more helpful than just this.

**Comment by ixrock** [comment]
I came up with such solution:

```javascript
// Utils.ts
export function convertEnumValuesToString(obj) {
  Object.keys(obj).forEach(function (key) {
    if (isNaN(+key)) {
      Object.defineProperty(obj, key, {
        value: key,
        enumerable: true
      });
    }
  });
  return obj;
}

// actions/hello.ts
import { convertEnumValuesToString } from '../utils'

export enum actionTypes {
  GREETING,
  FAREWELL_INIT,
  FAREWELL_DONE
}

console.log(actionTypes .GREETING) // 0
convertEnumValuesToString(actionTypes); // override to same key-based string values
console.log(actionTypes .GREETING) // "GREETING"
```

Maybe would be helpful to someone :)

**Comment by theRemix** [comment]
+1 would love to have this for angular2 ngrx reducer types.

i've been using this as string enums for now
```
export class Color {
  public static red = "red";
  public static blue = "blue";
  public static green = "green";
}
```

**Comment by amcdnl** [comment]
@theRemix check out this decorator I wrote - https://medium.com/@amcdnl/enums-in-es7-804a5a01bd70#.z0emdxdy1

**Comment by zspitz** [comment]
@dsherret Your hack doesn't work for ambient declarations.

**Comment by dsherret** [comment]
@zspitz just need to change the syntax accordingly:

```typescript
declare type MyStringEnum = "member1" | "member2";
declare const MyStringEnum: { Member1: MyStringEnum; Member2: MyStringEnum; };
```

**Comment by zspitz** [comment]
@dsherret This still won't inline the values (unlike ```const enum```), although I'm not sure any of the workarounds on this page will allow this, or if there is any language feature that supports inlining values other than ```number``` via enums.

**Comment by OliverJAsh** [comment]
TS 2 allows you to refer to enums as types as well as values, e.g.

``` ts
enum ColorTypes { Red, Green, Blue }
type RedColor = { type: ColorTypes.Red }
const redColor: RedColor = { type: ColorTypes.Red }
```

As far as I can see, none of the workarounds above will allow for this behaviour?

**Comment by RyanCavanaugh** [comment]
I think we can dupe this to #1206 since the only other type really being discussed there is `string` and there's some progress being made to some solution

**Comment by malcompm** [comment]
Hi TS Team, 
Seems like this is the most upvoted issue in this repository - so could we please at least get some update if this even get considered/done ? Seems like it does not even exist on the roadmap for Future https://github.com/Microsoft/TypeScript/wiki/Roadmap :(

**Comment by weswigham** [comment]
With `keyof` applied to a namespace or object, you can now trivially use an object or namespace of string literals as a "string enum". Given that enums aren't on the ECMAScript proposal table, and TS has avoided adding any further non-typesystem non-ECMAScript syntax for many moons, I would not hold my breath on this. Instead, I'd write this:

```ts
function mkenum<T extends {[index: string]: U}, U>(x: T) { return x; }
const Cardinal = mkenum({
  North: 'N',
  South: 'S',
  East: 'E',
  West: 'W'
});
type Cardinal = (typeof Cardinal)[keyof typeof Cardinal];
```
Flexible, powerful, relatively DRY. All TS could do with further syntax is imply the `type` line from extra syntax in the declaration and infer literal types by default which, again, is unlikely, given how little convince it offers over what is possible now, versus the extra risk if taking on new non-ECMAScript emit.

**Comment by wallverb** [comment]
@weswigham Thanks for this great idea - is there any way to adjust it so that it isn't using **widening** types ?

As soon as one does:
```ts
let test = Cardinal.North
```
the type of `test` becomes `string`

**Comment by weswigham** [comment]
@wallverb Use `const` or add a type annotation, otherwise, no. This is simply how string literal types are flowed now (the justification being: what is the use of assigning a single unit type to a mutable variable?).

**Comment by wallverb** [comment]
Or seems like one can do this (slightly less DRY but could prevent surprises and need to provide type annotations)

```ts
function mkenum<T extends {[index: string]: U}, U extends string>(x: T) { return x; }
const Cardinal = mkenum({
  North: 'N' as 'N',
  South: 'S' as 'S',
  East: 'E' as 'E',
  West: 'W' as 'W'
});
type Cardinal = (typeof Cardinal)[keyof typeof Cardinal];
```

and use it in cases like:
```ts
interface Go {
  cardinal: Cardinal
  miles: number
}
function go(where:Go){...}

const a = {
   cardinal: Cardinal.North,
   miles: 50
} // No need to annotate

go(a)  // <- otherwise we would get string not assignable to N|S|E|W due to type widening
```

For me that's a good enough alternative to string enums. Thanks a lot @weswigham / TS Team

**Comment by weswigham** [comment]
You don't need the 'mkenum' function when you explicitly write the
typecasts like that. The function call only existed to make the compiler do
exactly that for you. ;)

**Comment by mariusschulz** [comment]
> […] TS has avoided adding any further non-typesystem non-ECMAScript syntax for many moons.

While that is true, the required syntax is already in place in large parts. We already have const enums. Allowing strings as values (rather than only numbers) would "only" be a relaxation of a constraint that currently exists.

**Comment by OliverJAsh** [comment]
@weswigham Given your example, how come the final line here fails?

``` ts
function mkenum<T extends {[index: string]: U}, U extends string>(x: T) { return x; }
const Cardinal = mkenum({
  North: 'N',
  South: 'S',
  East: 'E',
  West: 'W'
});
type Cardinal = (typeof Cardinal)[keyof typeof Cardinal];
const p = Cardinal.East
const x = { y: Cardinal.East }
interface z { foo: Cardinal.East } // 'Cardinal' only refers to a type, but is being used as a value here.
```

**Comment by OliverJAsh** [comment]
Here's my workaround for string enums with type and value position support, until 2.1 lands with `keyof`:

``` ts
// Boilerplate for string enums until 2.1 (keyof)
// String enums can be used in type and value positions.
// https://github.com/Microsoft/TypeScript/issues/3192#issuecomment-261720275
export class ActionTypes {
	public static Foo: 'Foo' = 'Foo'
}

interface FooAction extends ReduxAction {
	type: typeof ActionTypes.Foo
}
```

**Comment by Kovensky** [comment]
@OliverJAsh The error in https://github.com/Microsoft/TypeScript/issues/3192#issuecomment-262187334 is correct. You _are_ using an "enum value" (`Cardinal.East`) as if it was a type. The `type Cardinal` is the union type of all valid `const Cardinal` values (in this example, `'N' | 'S' | 'E' | 'W'`).

If you want to define a type that has a key that can only ever have the value `Cardinal.East`, you're going to have to use `typeof Cardinal.East`; but that's probably not what you mean to do. `interface z { foo: Cardinal }` should work and allow any of the `Cardinal` values.

**Comment by nahuel** [comment]
@weswigham just a note about the `mkenum` example from https://github.com/Microsoft/TypeScript/issues/3192#issuecomment-261720275 : the inferred type here will be `string`, that's a different behaviour from using an standard numeric enum:

```ts
enum NumericEnum {
    A = 1
}

let a = NumericEnum.A              // a inferred type will be NumericEnum, good
let b = Cardinal.North             // b inferred type will be string, bad, is too generic, so ...
let c : Cardinal = Cardinal.North  // to have an useful type we must explicitely type c

```

**Comment by nahuel** [comment]
Here is a `mkenum` version were you don't need to specify the strings twice (merged with the previous mkenum example):

```ts
// original version:
function mkenum<T extends {[index: string]: U}, U extends string>(x: T) { return x; }
// new:
function mkenum2<X extends {[i:string] : any}, K extends string>(x : X ) 
  : {[K in (keyof X) ] : K}  {
    let o : any = {}
    for(let k in x) 
        o[k] = k;
    return o
}

// to define the type in a more succint way, for both versions:
export type enumType<T> = T[keyof T]

// define enums using the original and the new version:

const Colors  = mkenum({"Red"   : "Red", 
                        "Green" : "Green"});
type  Colors  = enumType<typeof Colors>

const Colors2 = mkenum2({"Red"   : 1, 
                         "Green" : 1});  // strings specified once, not twice
type  Colors2 = enumType<typeof Colors2>

let a  = Colors.Red       // string 
let a2 = Colors2.Red      // "Red"

let b  : Colors  = Colors.Red       // "Red" | "Green"
let b2 : Colors2 = Colors2.Red      // "Red" | "Green"
```

There is a little difference, the `Colors` constant is `{ Red : "Red", Green : "Green"}` but the `Colors2` constant is generated as `{ "Red" : "Red", "Green" : "Green"}`. 

Questions: 
1. What's generally better, the type inferred for `a` (`string`) or the one inferred for `a2` (the `"Red"` literal)? 
2. If the one inferred for `a` is better, how do you change `mkenum2` to generate a constant exactly like `Colors` to make `a2` be inferred with the same type as `a`? 
3. Is possible a `mkenum` version that takes as parameter an array like `["Red", "Green"]` ?

**Comment by Gambero81** [comment]
Now with typescript 2.1 and keyof, there is a better implementation of string enum? and what about online version for construction enum?

**Comment by zakjan** [comment]
```export type Enum = (typeof Enum)[keyof typeof Enum);``` works for me, even with no `mkenum` function

**Comment by igrayson** [comment]
@nahuel 
```ts
function mkenum3<X extends string>(...x:X[]):{[K in X]:K } {
  const o:any = {};
  for (const k in x)
    o[k] = k;
  return o;
}

type enumType<T> = T[keyof T];

const Colors3  = mkenum3('Red', 'Green');
type  Colors3  = enumType<typeof Colors3>;
```

**Comment by wallverb** [comment]
@igrayson Is there any way to modify your version so it behave the same as @nahuel `mkenum2` ?

```ts
let a2 = Colors2.Red // "Red"
let a3 = Colors3.Red // string -> would love to have "Red"
```

**Comment by dphilipson** [comment]
@igrayson Awesome solution! But I think you're loop is incorrect. It should be
```
function mkenum3<X extends string>(...x:X[]):{[K in X]:K } {
  const o:any = {};
  for (const k in x)
    o[x[k]] = x[k];
  return o;
}
```
that is, iterate over the values of the array not the keys.

When I ran your version, the types were correct, but the runtime value of `Colors3` was `["0": 0, "1", 1]`.

**Comment by dphilipson** [comment]
I use the techniques here in several of my projects, so I turned them into an NPM module for easy access. See https://github.com/dphilipson/typescript-string-enums.

I don't mean to claim credit for this solution, and I did my best to credit the users who came up with it in the readme. Hopefully others will find this useful.

**Comment by pelotom** [comment]
The [runtypes](https://github.com/pelotom/runtypes) library allows defining a runtime type for a union of literals and then extracting both the static type as well as its enumerated values. Example usage:

```ts
// Define the runtype
const Day = Union(
  Literal('Sunday'),
  Literal('Monday'),
  Literal('Tuesday'),
  Literal('Wednesday'),
  Literal('Thursday'),
  Literal('Friday'),
  Literal('Saturday'),
)

// Extract the static type
type Day = Static<typeof Day> // = 'Sunday' | 'Monday' | 'Tuesday' | 'Wednesday' | 'Thursday' | 'Friday' | 'Saturday'

// Extract enumerated literal values
const days: Day[] = Day.alternatives.map(lit => lit.value)

for (const day of days) {
  console.log(`Good morning, it's ${day}!`)
}
```

**Comment by kimamula** [comment]
With custom transformers introduced in #13940, you can create string enum from string literal types.

```
import { enumerate } from 'ts-transformer-enumerate';

type Colors = 'green' | 'yellow' | 'red';
const Colors = enumerate<Colors>(); // type of Colors is { [K in Colors]: K }

console.log(Colors.green); // 'green'
console.log(Colors.yellow); // 'yellow'
console.log(Colors.red); // 'red'
```

See https://github.com/kimamula/ts-transformer-enumerate.

**Comment by ahejlsberg** [comment]
Implementation now available in#15486.

**Comment by RyanCavanaugh** [comment]
Let's track this at #1206
